module/zfs/dmu.c

```diff
@@ -474,20 +474,23 @@ dmu_buf_hold_array_by_dnode(dnode_t *dn, uint64_t offset, uint64_t length,
 	}
 	dbp = kmem_zalloc(sizeof (dmu_buf_t *) * nblks, KM_SLEEP);
 
-	zio = zio_root(dn->dn_objset->os_spa, NULL, NULL, ZIO_FLAG_CANFAIL);
```

Comment:
I think it would be easier to read this code if we always created the zio for when `read` is set.  We should get all the benefit of your change in the common case.  And only in the rare error case where the `dbuf_hold()` fails and we return `EIO` would be incur the allocation/nowait cost.  Then the changes on lines 483 and 489 wouldn't be needed.

```diff
-       zio = zio_root(dn->dn_objset->os_spa, NULL, NULL, ZIO_FLAG_CANFAIL);
+       if (read) {
+               zio = zio_root(dn->dn_objset->os_spa, NULL, NULL,
+                   ZIO_FLAG_CANFAIL);
+       }
+
```